### ISSUE TITLE
imap_perform_authentication() does not seem to fallback to cleartext authentication as expected.

### ISSUE THREAD

**Issue by kop316** [curl/curl#8984]
Hello! For a bit of background, I developed and maintain an application called vvmd (a Visual Voicemail application for mobile phones).

Visual Voicemail works primarily over IMAP, so I use libcurl as my backend.

### I did this

Recently, we discovered that IMAP authentication was not working for a subset of T-Mobile USAs IMAP servers:
https://gitlab.com/kop316/vvmd/-/issues/12

Mobile carriers are notorious for misconfigured IMAP servers, so in `vvmd`, I do a variety of login tests to determine the correct authentication method for an IMAP Voicemail server.

In T-mobile's case, the IMAP server can advertise
```
< * OK [CAPABILITY AUTH=PLAIN AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
> A001 CAPABILITY
< * CAPABILITY AUTH=PLAIN AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
```

But attempts to authenticate with `AUTH=DIGEST-MD5` do not work with T-Mobile's IMAP server, only `AUTH=PLAIN`:

```
> A001 CAPABILITY
< * CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
< A001 OK CAPABILITY completed.
> A002 AUTHENTICATE DIGEST-MD5
< + cmVhbG09InJlYWxtQG1hdmVuaXIuY29tIixub25jZT0iTURRNVptWmhOek10TVRZek5DMDBNVEppTFRrMVlXRXRNelpoWldKak16ZGlaR1ZqIixxb3A9ImF1dGgiLGNoYXJzZXQ9dXRmLTgsYWxnb3JpdGhtPW1kNS1zZXNz
> dXNlcm5hbWU9IjE0NDM2ODM5MTE5IixyZWFsbT0icmVhbG1AbWF2ZW5pci5jb20iLG5vbmNlPSJNRFE1Wm1aaE56TXRNVFl6TkMwME1USmlMVGsxWVdFdE16WmhaV0pqTXpkaVpHVmoiLGNub25jZT0iNWUyZTYxMjI2MzJjMjU5ODVkZTc1ZDJiYzI4M2IyODYiLG5jPSIwMDAwMDAwMSIsZGlnZXN0LXVyaT0iaW1hcC9yZWFsbUBtYXZlbmlyLmNvbSIscmVzcG9uc2U9ZGFlZDMwODg1MzllODcyZGRiMWZkZmIyYjJlMGY2YzgscW9wPWF1dGg=
< A002 NO application error
```

I determined that the reason that IMAP authentication no longer occured for the subset of IMAP servers was because those servers no longer advertise `AUTH=PLAIN`

```
< * OK [CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
> A001 CAPABILITY
< * CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
```

Since `AUTH=PLAIN` is no longer an option, curl only attempts to authenticate with `AUTH=DIGEST-MD5`, which, as stated before, does not work.

However, as shown below, if I attempt plain authentication manually, I get a successful login:

```
$openssl s_client -crlf -connect e4.vvm.mstore.msg.t-mobile.com:993
...
<Truncated output>
...
* OK [CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
a01 LOGIN $USERNAME $PASSWORD
a01 OK LOGIN completed.
a02 LIST "" *
* LIST (\HasNoChildren) "/" "/INBOX"
* LIST (\HasNoChildren) "/" "/Greetings"
* LIST (\HasNoChildren) "/" "/Trash"
a02 OK LIST completed.
* BYE Autologout; idle for too long
read:errno=104
```
I additionally attempted:
```
curl -v --login-options AUTH=PLAIN imaps://$USERNAME:$PASSWORD@e4.vvm.mstore.msg.t-mobile.com/
```
However, I get the following:
```
< * OK [CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
> A001 CAPABILITY
< * CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
< A001 OK CAPABILITY completed.
* No known authentication mechanisms supported
* Closing connection 0
* TLSv1.2 (OUT), TLS alert, close notify (256):
curl: (67) Login denied
```
For completeness, here are the full logs: https://pastebin.com/1BjEYdbH

Unfortunately, I have no means to contact T-Mobile and have this fixed.

### I expected the following

The `No known authentication mechanisms supported` looks to track here:
https://github.com/curl/curl/blob/master/lib/imap.c#L601

`imap_perform_authentication()` has a comment of 

```
falling back to clear text should a common mechanism not be available between the client and server.
```

so I was expecting authentication to fall back to plaintext, but it looks like `IMAP_TYPE_CLEARTEXT` is `FALSE`, (since there is no `AUTH=PLAIN`) so curl never attempts to `imap_perform_login()`

If that comment is outdated and this is expected behavior, it would be nice if:

```
curl -v --login-options AUTH=PLAIN imaps://$USERNAME:$PASSWORD@e4.vvm.mstore.msg.t-mobile.com/
```

Attempts `AUTH=PLAIN` regardless as to whether the IMAP server advertises the capability or not, or if there could be an additional flag to tell curl that I don't care what the IMAP server advertises, do as I say.

Additonal for libcurl, it would be nice to have something analagous for `CURLOPT_LOGIN_OPTIONS` to force a certain login option regardless of what the server advertises.

### curl/libcurl version

```
curl 7.74.0 (x86_64-pc-linux-gnu) libcurl/7.74.0 OpenSSL/1.1.1n zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libpsl/0.21.0 (+libidn2/2.3.0) libssh2/1.9.0 nghttp2/1.43.0 librtmp/2.3
Release-Date: 2020-12-09
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

### operating system

I am using Debian Bullseye

```
$uname -a
Linux Spaceballsthex2100 5.10.0-14-amd64 #1 SMP Debian 5.10.113-1 (2022-04-29) x86_64 GNU/Linux
```

This reliably also happens in Debian Bookworm, PureOS, and postmarketOS

EDIT: Sorry, I used `openssl s_client -crlf -connect e4.vvm.mstore.msg.t-mobile.com:993` for manual authentication, not curl

**Comment by captain-caveman2k** [comment]
I see you have found #8979 where I have a proposed fix to allow fallback to the LOGIN command after a failed AUTHENTICATE command - I have since extended that fix to include a new test case which I have in my own folk of curl at the moment. I hope to publish this soon and run it through the CI process ;-)

FYI: Using AUTH=PLAIN is not the same as performing an IMAP authentication via the LOGIN command. Whilst the LOGIN command transmits authentication details in plain/clear text is should not be confused with the SASL based authentication mechanisms PLAIN and LOGIN that can be performed via the AUTHENTICATE command.

**Comment by kop316** [comment]
> I see you have found #8979 where I have a proposed fix to allow fallback to the LOGIN command after a failed AUTHENTICATE command - I have since extended that fix to include a new test case which I have in my own folk of curl at the moment. I hope to publish this soon and run it through the CI process ;-)

That is great! Please let me know, I would be happy to test it out and see if it works (as I assume you mean the fix can cover this issue as well). 

> 
> FYI: Using AUTH=PLAIN is not the same as performing an IMAP authentication via the LOGIN command. Whilst the LOGIN command transmits authentication details in plain/clear text is should not be confused with the SASL based authentication mechanisms PLAIN and LOGIN that can be performed via the AUTHENTICATE command.

Ohh ok, sorry for the confusion! I tried looking up the RFC to see if I could find if there is a known fallback for `AUTH=PLAIN`....but I found that there is a lot of RFCs that cover IMAP, so I could not find the one governing this. If you happen to know it, I would very much appreciate a link!

**Comment by captain-caveman2k** [comment]
> That is great! Please let me know, I would be happy to test it out and see if it works (as I assume you mean the fix can cover this issue as well).

Yes the fix should cover both scenarios as both need to fallback to the LOGIN command. Ironically the title of this issue is more fitting for my fix ;-)

That would be helpful as I'm in San Francisco at the moment and the WSL installation on my laptop is playing up :(

If you are familiar with git, have a copy of the curl repo and fancy grabbing the diff from the other issue then simply save it to a .diff file and use git apply to make the change. Then recompile the curl source and try it out ;-)

> Ohh ok, sorry for the confusion!

No worries - It took me a little whilst yesterday when I was starting to look the issue and I have a number of years experience with IMAP and specifically curl's implementation ;-)

IMAP is also a little odd IMO, in that it has two commands for authentication, and that the second command does something that can be covered by a cleartext SASL mechanism (PLAIN or LOGIN) in the first, albeit, the format (in the case of PLAIN) or transmission (in the case of LOGIN) are different.

It also doesn't help in that we don't support specifying only using the LOGIN command when SASL authentication mechanisms are present in the CAPABILITIES list via our login options. As AUTH=<mechanism> is used to specify a SASL authentication mechanism I guess (and I'm totally thinking off the top of my head) we could extend that in IMAP to support LOGIN which might look something like the following when specified in the url:

`imap://user:password@imap.example.com;LOGIN`

as opposed to:

- `imap://user:password@imap.example.com;AUTH=PLAIN`, or
- `imap://user:password@imap.example.com;AUTH=LOGIN`

>  I tried looking up the RFC to see if I could find if there is a known fallback for AUTH=PLAIN....but I found that there is a lot of RFCs that cover IMAP, so I could not find the one governing this. If you happen to know it, I would very much appreciate a link!

The main RFCs that I know of are:

- [RFC-1730 (IMAP 4)](https://www.ietf.org/rfc/rfc1730.html)
- [RFC-2060 (IMAP 4rev1)](https://www.ietf.org/rfc/rfc2060.html)
- [RFC-3501 (IMAP 4rev1 update)](https://www.ietf.org/rfc/rfc3501.html) - This is the one that curl has followed for years
- [RFC-9051 (IMAP 4rev2)](https://www.ietf.org/rfc/rfc9051.html) - This is the recent revision that explicitly mentions the fallback to the LOGIN command

There are also a number of updates and extensions to most of the RFCs - these are mention in the updates section at the top of each.

The following resource is also useful for finding out what CAPABILITIES where added in what RFCs:

[IMAP Capabilities](https://www.iana.org/assignments/imap-capabilities/imap-capabilities.xhtml)

**Comment by kop316** [comment]
> Yes the fix should cover both scenarios as both need to fallback to the LOGIN command. Ironically the title of this issue is more fitting for my fix ;-)
> That would be helpful as I'm in San Francisco at the moment and the WSL installation on my laptop is playing up :(
> If you are familiar with git, have a copy of the curl repo and fancy grabbing the diff from the other issue then simply save it to a .diff file and use git apply to make the change. Then recompile the curl source and try it out ;-)

I know enough that I could fumble my way through the process (assuming curl isn't too hard to compile, which I doubt it is). I will try it out tonight and let you know what ends up happening. 

> IMAP is also a little odd IMO, in that it has two commands for authentication, and that the second command does something that can be covered by a cleartext SASL mechanism (PLAIN or LOGIN) in the first, albeit, the format (in the case of PLAIN) or transmission (in the case of LOGIN) are different.
> It also doesn't help in that we don't support specifying only using the LOGIN command when SASL authentication mechanisms are present in the CAPABILITIES list via our login options. As AUTH= is used to specify a SASL authentication mechanism I guess (and I'm totally thinking off the top of my head) we could extend that in IMAP to support LOGIN which might look something like the following when specified in the URL.

> The main RFCs that I know of are

Thanks! I am curious to see if `AUTH=PLAIN` has a similar fallback mechanism, but as you commented, it is quite a bit confusing.

**Comment by captain-caveman2k** [comment]
> I know enough that I could fumble my way through the process (assuming curl isn't too hard to compile, which I doubt it is). I will try it out tonight and let you know what ends up happening.

Cheers - that is appreciated if you can. Alternatively I can try and sort WSL out or even VPN onto my desktop or the curl build servers.

> Thanks! I am curious to see if `AUTH=PLAIN` has a similar fallback mechanism, but as you commented, it is quite a bit confusing.

Note: You probably won't find mention of AUTH=PLAIN as that is a URL extension and curl login option. You should find the AUTHENTICATE command described in section 6.2.2 of RFC-9051 (for example) which is what is used to authenticate with any of the SASL authentication mechanisms (PLAIN, LOGIN, NTLM, GSS-API, CRAM-MD5, EXTERNAL, etc...).

You will find the LOGIN command described in the next section (6.2.3) where it quotes the paragraph about fallback. You'll see the fallback applies to the failure of the AUTHENTICATE command rather than any specific SASL mechanism.

**Comment by kop316** [comment]
@captain-caveman2k 

I can confirm that the patch worked!

(Note that `$USERNAME` is my username `$PASSWORD` is my password)

When I did:
```
$ ./src/curl -v imaps://$USERNAME:$PASSWORD@e4.vvm.mstore.msg.t-mobile.com/
```

```
* STATE: INIT => CONNECT handle 0x5574b1be5088; line 1835 (connection #-5000)
* Added connection 0. The cache now contains 1 members
* STATE: CONNECT => RESOLVING handle 0x5574b1be5088; line 1881 (connection #0)
* family0 == v4, family1 == v6
*   Trying 208.54.154.75:993...
* STATE: RESOLVING => CONNECTING handle 0x5574b1be5088; line 1965 (connection #0)
* Connected to e4.vvm.mstore.msg.t-mobile.com (208.54.154.75) port 993 (#0)
* STATE: CONNECTING => PROTOCONNECT handle 0x5574b1be5088; line 2028 (connection #0)
* IMAP 0x5574b1bddd50 state change from STOP to SERVERGREET
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* Didn't find Session ID in cache for host IMAPS://e4.vvm.mstore.msg.t-mobile.com:993
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* STATE: PROTOCONNECT => PROTOCONNECTING handle 0x5574b1be5088; line 2048 (connection #0)
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / AES256-GCM-SHA384
* Server certificate:
*  subject: C=US; ST=Washington; L=Bellevue; O=T-Mobile USA, Inc.; CN=vvm.mstore.msg.t-mobile.com
*  start date: Sep 22 19:14:55 2021 GMT
*  expire date: Sep 22 19:14:55 2022 GMT
*  subjectAltName: host "e4.vvm.mstore.msg.t-mobile.com" matched cert's "e4.vvm.mstore.msg.t-mobile.com"
*  issuer: C=US; O=Entrust, Inc.; OU=See www.entrust.net/legal-terms; OU=(c) 2012 Entrust, Inc. - for authorized use only; CN=Entrust Certification Authority - L1K
*  SSL certificate verify ok.
< * OK [CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
> A001 CAPABILITY
* IMAP 0x5574b1bddd50 state change from SERVERGREET to CAPABILITY
< * CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
< A001 OK CAPABILITY completed.
> A002 AUTHENTICATE DIGEST-MD5
* SASL 0x5574b1bddde8 state change from STOP to DIGESTMD5
* IMAP 0x5574b1bddd50 state change from CAPABILITY to AUTHENTICATE
< + cmVhbG09InJlYWxtQG1hdmVuaXIuY29tIixub25jZT0iTVdReU5UQmhZamd0TURaa09TMDBaREU1TFRsak5tSXRPVEkzTm1Rd09XWmhNamRtIixxb3A9ImF1dGgiLGNoYXJzZXQ9dXRmLTgsYWxnb3JpdGhtPW1kNS1zZXNz
> dXNlcm5hbWU9IjE0NDM2ODM5MTE5IixyZWFsbT0icmVhbG1AbWF2ZW5pci5jb20iLG5vbmNlPSJNV1F5TlRCaFlqZ3RNRFprT1MwMFpERTVMVGxqTm1JdE9USTNObVF3T1daaE1qZG0iLGNub25jZT0iNDI0YTkxNjk2NTUwZmJmZjc3YjUyM2I5NzhlZTk2YWUiLG5jPSIwMDAwMDAwMSIsZGlnZXN0LXVyaT0iaW1hcC9yZWFsbUBtYXZlbmlyLmNvbSIscmVzcG9uc2U9NWExNDhmMjYxNzc2MzY2ZjdjZTI0MjYzM2YyMmM2YWQscW9wPWF1dGg=
* SASL 0x5574b1bddde8 state change from DIGESTMD5 to DIGESTMD5_RESP
< A002 NO application error
* SASL 0x5574b1bddde8 state change from DIGESTMD5_RESP to STOP
> A003 LOGIN $USERNAME $PASSWORD
* IMAP 0x5574b1bddd50 state change from AUTHENTICATE to LOGIN
< A003 OK LOGIN completed.
* IMAP 0x5574b1bddd50 state change from LOGIN to STOP
* STATE: PROTOCONNECTING => DO handle 0x5574b1be5088; line 2067 (connection #0)
* DO phase starts
> A004 LIST "" *
* IMAP 0x5574b1bddd50 state change from STOP to LIST
* STATE: DO => DOING handle 0x5574b1be5088; line 2134 (connection #0)
< * LIST (\HasNoChildren) "/" "/INBOX"
* LIST (\HasNoChildren) "/" "/INBOX"
< * LIST (\HasNoChildren) "/" "/Greetings"
* Curl_pp_readresp_ 63 bytes of trailing server response left
* LIST (\HasNoChildren) "/" "/Greetings"
< * LIST (\HasNoChildren) "/" "/Trash"
* Curl_pp_readresp_ 25 bytes of trailing server response left
* LIST (\HasNoChildren) "/" "/Trash"
< A004 OK LIST completed.
* IMAP 0x5574b1bddd50 state change from LIST to STOP
* DO phase is complete
* STATE: DOING => DID handle 0x5574b1be5088; line 2216 (connection #0)
* STATE: DID => PERFORMING handle 0x5574b1be5088; line 2266 (connection #0)
* STATE: PERFORMING => DONE handle 0x5574b1be5088; line 2465 (connection #0)
* multi_done: status: 0 prem: 0 done: 0
* Connection #0 to host e4.vvm.mstore.msg.t-mobile.com left intact
* Expire cleared (transfer 0x5574b1be5088)
```
Something I noted though:

when I did:
```
$ ./src/curl -v --login-options AUTH=PLAIN imaps://14436839119:8xYI7fQbzg@e4.vvm.mstore.msg.t-mobile.com/
```

I still got:
```
*  SSL certificate verify ok.
< * OK [CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
> A001 CAPABILITY
* IMAP 0x55a2c0a40d50 state change from SERVERGREET to CAPABILITY
< * CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
< A001 OK CAPABILITY completed.
* No known authentication mechanisms supported
* multi_done: status: 67 prem: 1 done: 0
* The cache now contains 0 members
* Closing connection 0
* TLSv1.2 (OUT), TLS alert, close notify (256):
* Expire cleared (transfer 0x55a2c0a48088)
```

Which I suppose isn't too surprising (since it never has anything to fallback to).

I also got:
```
$ ./src/curl -v --login-options AUTH=DIGEST-MD5 imaps://14436839119:8xYI7fQbzg@e4.vvm.mstore.msg.t-mobile.com/
```

```
< * OK [CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS] mStore IMAP4rev1 Server is ready.
> A001 CAPABILITY
* IMAP 0x55866cd75d50 state change from SERVERGREET to CAPABILITY
< * CAPABILITY AUTH=DIGEST-MD5 IMAP4rev1 QUOTA UIDPLUS
< A001 OK CAPABILITY completed.
> A002 AUTHENTICATE DIGEST-MD5
* SASL 0x55866cd75de8 state change from STOP to DIGESTMD5
* IMAP 0x55866cd75d50 state change from CAPABILITY to AUTHENTICATE
< + cmVhbG09InJlYWxtQG1hdmVuaXIuY29tIixub25jZT0iT0dRMU1XSmlZamt0TVdFd1lTMDBORGhpTFRnd01tUXRORFU1WlRkak16UTVNbU5sIixxb3A9ImF1dGgiLGNoYXJzZXQ9dXRmLTgsYWxnb3JpdGhtPW1kNS1zZXNz
> dXNlcm5hbWU9IjE0NDM2ODM5MTE5IixyZWFsbT0icmVhbG1AbWF2ZW5pci5jb20iLG5vbmNlPSJPR1ExTVdKaVlqa3RNV0V3WVMwME5EaGlMVGd3TW1RdE5EVTVaVGRqTXpRNU1tTmwiLGNub25jZT0iOTlhN2I5OGI1OTg4MmE3YmRkMzIxZDg1NGEzOTAzNzgiLG5jPSIwMDAwMDAwMSIsZGlnZXN0LXVyaT0iaW1hcC9yZWFsbUBtYXZlbmlyLmNvbSIscmVzcG9uc2U9MWViMjgxMzNmMjVjNTE4NWY3NmZiMDQwZDhkMmE3MGIscW9wPWF1dGg=
* SASL 0x55866cd75de8 state change from DIGESTMD5 to DIGESTMD5_RESP
< A002 NO application error
* SASL 0x55866cd75de8 state change from DIGESTMD5_RESP to STOP
* multi_done: status: 67 prem: 1 done: 0
* The cache now contains 0 members
* Closing connection 0
* TLSv1.2 (OUT), TLS alert, close notify (256):
* Expire cleared (transfer 0x55866cd7d088)
curl: (67) Login denied
```

I was sort of surprised to see that curl didn't decide to fall back like it did before. I am not sure if that is unexpected or undesired behavior, but I wanted to let you know.

I also did not specifcally try this with `vvmd` (I don't know how to integrate this into `vvmd`), but I assume if it works with curl, it will likewise work for libcurl.

**Comment by captain-caveman2k** [comment]
> I can confirm that the patch worked!

Cool - thank you. I accessed my development box at home, as well, and proved that the new test case ran as expected too.

> Something I noted though:
> 
> ```
> $ ./src/curl -v --login-options AUTH=PLAIN imaps://14436839119:8xYI7fQbzg@e4.vvm.mstore.msg.t-mobile.com/
> ```

As you mentioned curl hasn't got anything to fallback to as you have specified a specific mechanism via the login options.

If we were to support LOGIN command by adding it to the login options then I would expect the previous behaviour. I believe you can specify multiple options (it's a long time since I wrote that code, and it has changed since as well) but then you could specify the following for example:

AUTH=DIGEST-MD5;LOGIN

> 
> ```
> $ ./src/curl -v --login-options AUTH=DIGEST-MD5 imaps://14436839119:8xYI7fQbzg@e4.vvm.mstore.msg.t-mobile.com/
> ```
>
> I was sort of surprised to see that curl didn't decide to fall back like it did before. I am not sure if that is unexpected or undesired behavior, but I wanted to let you know.

Did you mean to pass in the login options there?

> I also did not specifcally try this with `vvmd` (I don't know how to integrate this into `vvmd`), but I assume if it works with curl, it will likewise work for libcurl.

When you built for your desktop you would have generated a new libcurl as well, however, you should be able to build libcurl for your mobile platform as well (I'm assuming vvmd is Android based) as there are some Android files under packages\Android.

Whilst I used to be involved in the mobile industry between 2002 and 2010 I have never built curl for mobile unfortunately :(

**Comment by kop316** [comment]
> Did you mean to pass in the login options there?

Ack! no. I editted my comment, please edit yours as well.

**Comment by kop316** [comment]
> If we were to support the LOGIN command by adding it to the login options then I would expect the previous behaviour. I believe you can specify multiple options (it's a long time since I wrote that code, and it has changed since as well) but then you could specify the following for example:

> AUTH=DIGEST-MD5;LOGIN

Ahh, yes that makes sense! I couldn't figure out the delimiter, but `AUTH=*` does work with the patch.

> When you built for your desktop you would have generated a new libcurl as well, however, you should be able to build libcurl for your mobile platform as well (I'm assuming vvmd is Android based) as there are some Android files under packages\Android.

Fair enough! I will see if I can figure this out, but I am expecting it to work.

On a side note: This is actually all being done on a Pinephone, so it is with GNU/Linux based Distribution called [Mobian](https://mobian-project.org/). It is a distro that has a mobile overlay with Debian (but with the long term goal of merging into Debian Proper).

**Comment by captain-caveman2k** [comment]
> On a side note: This is actually all being done on a Pinephone, so it is with GNU/Linux based Distribution called [Mobian](https://mobian-project.org/). It is a distro that has a mobile overlay with Debian (but with the long term goal of merging into Debian Proper).

Oh - that sounds complicated. I used to program for Symbian, Microsoft Pocket PC and Smartphone, as well as some embedded C phones back in the naughties but I haven't touched anything post Apple iPhone really.

**Comment by kop316** [comment]
Hello!

For completeness, I did test this with `vvmd`, and I can comfirn the fix here: https://github.com/curl/curl/issues/8979#issuecomment-1150232613 worked.

**Comment by fuzzy7k** [comment]
ping @captain-caveman2k

I see that your PR was closed. 

As it conflicted with https://github.com/curl/curl/pull/8295, which is massive compared to your pull request, I wondered if work stalled because you were waiting on that to be ironed out.

I'll +1 vote for this because it means functioning voicemail for my parents.

**Comment by kop316** [comment]
Hello!

To add to @fuzzy7k 's vote, I would like to vote +1 as well. For some background, I have been the primary developer for Visual Voicemail on Mobile Linux Distributions (e.g. on the Pinephone/Pinephone Pro/ Librem 5). 

Some carriers rely on this behavior to properly access their IMAP servers. The fix that was just closed fixed the behavior, and I would like to understand what I can do to help, whether that involves paying for dev time or reworking the PR myself.

**Comment by rothn** [comment]
This affects me and breaks my voicemail. Any ETA on a release to Manjaro?

**Comment by rothn** [comment]
Do you think you could patch this upstream while we wait so our phones can work?

**Comment by kop316** [comment]
This can be closed with https://github.com/curl/curl/commit/64aefea3d9aca43ea84a7ad8a606f4ee49cb1eab

Thanks!
